File: kimssumin/Q33.py
def solution(triangle):
    answer = 0
    nt = [[0] + t + [0] for t in triangle]
    
    for i in range(1, len(nt)):
        for j in range(1, i+2):
            nt[i][j] += max(nt[i-1][j-1], nt[i-1][j])
    
    return max(nt[-1])

File: BaeYoungSuk/Q33.java
import java.util.*;

class Solution {
    public int solution(int[][] triangle) {
        for(int i = 0; i < triangle.length - 1; i++){
            int[] upper = triangle[i];
            int[] cur = triangle[i+1];
            for(int j = 0; j < cur.length; j++){
                if(j == 0) {
                    cur[j] += upper[0];
                } else if(j == cur.length - 1){
                    cur[j] += upper[upper.length - 1];
                } else {
                    cur[j] += Math.max(upper[j-1], upper[j]);
                }
            }
            triangle[i+1] = cur;
        }
        int max = Integer.MIN_VALUE;
        int[] arr = triangle[triangle.length - 1];
        for(int e : arr) {
            if(e > max){
                max = e;
            }
        }
        return max;
    }
}
